io-classes/src/Control/Monad/Class/MonadUnique.hs

```diff
@@ -0,0 +1,33 @@
+{-# LANGUAGE TypeFamilies, AllowAmbiguousTypes #-}
+
+-- | A generalisation of the
+-- <https://hackage.haskell.org/package/base/docs/Data-Unique.html Data.Unique>
+-- API to both 'IO' and <https://hackage.haskell.org/package/io-sim IOSim>.
+--
+module Control.Monad.Class.MonadUnique (
+  MonadUnique(..)
+) where
+
+-- base
+import Data.Kind (Type)
+import Data.Unique qualified as IO
+
+-- transformers
+import Control.Monad.Reader (ReaderT(..), lift)
+
+
+class (Monad m, Eq (Unique m), Ord (Unique m)) => MonadUnique m where
+  type Unique m :: Type
+  newUnique  :: m (Unique m)
+  hashUnique :: Unique m -> Int
+
+instance MonadUnique IO where
+  type Unique IO = IO.Unique
+  newUnique  = IO.newUnique
+  hashUnique = IO.hashUnique
+
+instance MonadUnique m => MonadUnique (ReaderT r m) where
+  type Unique (ReaderT r m) = Unique m
+  newUnique  = lift newUnique
+  hashUnique = hashUnique @m
```

Comment:
Could you also add instances for other mtl type classes in `io-classes:mtl` sublibrary?

Reply:
Done.